ee/clickhouse/queries/funnels/base.py

```diff
@@ -102,11 +102,21 @@ def _get_steps_per_person_query(self):
             formatted_query = self._get_inner_event_query()
 
         return f"""
-        SELECT *, {self._get_sorting_condition(max_steps, max_steps)} AS steps FROM (
+        SELECT *, {self._get_sorting_condition(max_steps, max_steps)} AS steps {self._get_step_times(max_steps)} FROM (
             {formatted_query}
         ) WHERE step_0 = 1
         """
 
+    def _get_step_times(self, max_steps: int):
+        conditions: List[str] = []
+        for i in range(1, max_steps):
+            conditions.append(
+                f"if(isNotNull(latest_{i}), dateDiff('second', toDateTime(latest_{i - 1}), toDateTime(latest_{i})), NULL) step_{i}_average_conversion_time"
+            )
+
+        formatted = ", ".join(conditions)
+        return f", {formatted}" if formatted else ""
```

Comment:
A question about the future / how we build queries in Python: I'd really like us to get to a standard, such that either:

(1) Every function assumes there's no commas, and adds them itself. (what `_get_step_times` is doing)
or
(2) Everything is setup in the outer query building, and specific functions just return what the specific query bit is.

With my limited interaction in the past week with this codebase, I _feel_ (2) makes more sense, keeps things reusable and clean, while making it a bit more complicated to write the outer query (conditionals to insert commas). This is a reasonable tradeoff imo.

Moving forward, I'd suspect we'd see more sub-query-builder-functions like these, and I _think_ it becomes unfeasible to keep in mind where all the commas go 😅 .

---

Thoughts?

Reply:
Yeah, outer query handling makes sense to me especially as sometimes the outer query will have to ensure the select doesn't start with a comma accidentally